README
               LMDBG - lightweight malloc debugger

====================================================================
                      INSTALLATION

See INSTALL file

====================================================================
                      DOCUMENTATION/TUTORIAL

- LMDBG is a lightweight malloc debugger for UNIX-like OSes that
  follows so called "UNIX way", traditional principles of programming
  under UNIX. Instead of providing single program that can do
  everything LMDBG provides a set of standalone programs that work
  together: lmdbg-run, lmdbg-sym, lmdbg-leaks, lmdbg-sysleaks,
  lmdbg-stat, lmdbg-grep and lmdbg-sort.

- Unlike many other malloc debuggers (electric-fence, dmalloc,
  valgrind etc.) LMDBG does *NOT* protect your memory in any way.
  This is just not a goal of it.

- lmdbg-run is a main lmdbg utility. It runs application and creates a
  log file (or fifo) where all
  malloc/calloc/realloc/free/memalign/posix_memalign invocations are
  registered with their input (bytes count and/or pointer), output
  (pointer) and FULL STACKTRACE (pointers).

   Example:
     $ cat tests/test2.c
     #include <stdlib.h>

     int main ()
     {
        void *p1 = NULL;
        void *p2 = NULL;

        p1 = malloc (555);
        p2 = realloc (p2, 666);
        p2 = realloc (p2, 777);
        p2 = realloc (p2, 888);

        return 0;
     }
     $ gcc -O0 -g -o _test2 tests/test2.c
     $ lmdbg-run -o _log ./_test2 
     $ cat _log
     malloc ( 555 ) --> 0xbb901400
      0xbbbe58e8
      0xbbbe5b03
      0x8048738
      0x8048584
      0x80484e7
     realloc ( NULL , 666 ) --> 0xbb901800
      0xbbbe58e8
      0xbbbe5a37
      0x804874e
      0x8048584
      0x80484e7
     realloc ( 0xbb901800 , 777 ) --> 0xbb901c00
      0xbbbe58e8
      0xbbbe5a37
      0x8048764
      0x8048584
      0x80484e7
     realloc ( 0xbb901c00 , 888 ) --> 0xbb901800
      0xbbbe58e8
      0xbbbe5a37
      0x804877a
      0x8048584
      0x80484e7
     $

  NOTE: Full stacktrace allows you to analyse your application, i.e.
  you can detect what blocks/components require more memory than
  others and why. lmdbg-sym is a very important tool for this, see
  below.

- lmdbg-leaks analyses a log file generated by lmdbg-run and output
  all found memory leaks

  Example:

     $ lmdbg-leaks _log   
     realloc ( 0xbb901c00 , 888 ) --> 0xbb901800
      0xbbbe58e8
      0xbbbe5a37
      0x804877a
      0x8048584
      0x80484e7
     malloc ( 555 ) --> 0xbb901400
      0xbbbe58e8
      0xbbbe5b03
      0x8048738
      0x8048584
      0x80484e7
     $

- lmdbg-sym converts addresses to symbols and position in the source
  code if it is possible

  Example:

     $ lmdbg-sym ./_test2 _log
     malloc ( 555 ) --> 0xbb901400
      0xbbbe58e8
      0xbbbe5b03
      0x8048738      tests/test2.c:8    main
      0x8048584
      0x80484e7
     realloc ( NULL , 666 ) --> 0xbb901800
      0xbbbe58e8
      0xbbbe5a37
      0x804874e      tests/test2.c:9    main
      0x8048584
      0x80484e7
     realloc ( 0xbb901800 , 777 ) --> 0xbb901c00
      0xbbbe58e8
      0xbbbe5a37
      0x8048764      tests/test2.c:10        main
      0x8048584
      0x80484e7
     realloc ( 0xbb901c00 , 888 ) --> 0xbb901800
      0xbbbe58e8
      0xbbbe5a37
      0x804877a      tests/test2.c:11        main
      0x8048584
      0x80484e7
     $

- lmdbg-sysleaks - greps or skips system memory leaks found in
  libc, libdl, C++ stl etc. See tests/lmdbg*.conf files.
  The default config files are: ~/.lmdbg.conf and /etc/lmdbg.conf

- lmdbg = lmdbg-run + lmdbg-leaks + lmdbg-sym + lmdbg-sysleaks

  That is lmdbg is all-in-one higher level tool.

   Example:

     $ lmdbg -v -o _log ./_test2
     Memory leaks were detected and saved to file '_log'
     $ cat _log
     realloc ( 0xbb901c00 , 888 ) --> 0xbb901800
      0xbbbe58e8
      0xbbbe5a37
      0x804877a      tests/test2.c:11        main
      0x8048584
      0x80484e7
     malloc ( 555 ) --> 0xbb901400
      0xbbbe58e8
      0xbbbe5b03
      0x8048738      tests/test2.c:8 main
      0x8048584
      0x80484e7
     $

- lmdbg-stat - given output of lmdbg-run or other lmdbg-* utilities
  on input lmdbg-stat outputs a total and per-stacktrace statistical
  information about memory alocations.

     $ cat -n testme.c
          1  #include <stdlib.h>
          2  
          3  int main ()
          4  {
          5          int i;
          6          void *p1, *p2;
          7          p1 = malloc (100);
          8          for (i=0; i < 3; ++i){
          9                  p2 = realloc (NULL, 10+30*i);
          10          }
          11          free (p1);
          12          free (p2);
          13  
          14          return 0;
          15  }
     $ cc -O0 -g -o testme testme.c
     $ lmdbg-run -o log -p 'lmdbg-stat | lmdbg-sym ./testme' ./testme     
     $ cat log
     info stat total_leaks: 50
     info stat total_allocs: 4
     info stat total_free_cnt: 2
     stacktrace peak: 120 max: 70 allocs: 3 leaks: 50
      0xbbbe2bc3     lmdbg.c:101     log_stacktrace
      0xbbbe3498     lmdbg.c:456     realloc
      0x8048788      testme.c:9      main
      0x80485b4      
      0x8048517      
     stacktrace peak: 100 max: 100 allocs: 1
      0xbbbe2bc3     lmdbg.c:101     log_stacktrace
      0xbbbe33bd     lmdbg.c:431     malloc
      0x8048757      testme.c:7      main
      0x80485b4      
      0x8048517      
     $ 

- lmdbg-grep - given an output of lmdbg-stat on input lmdbg-grep
  outputs global information and those stacktraces that match the
  specified AWK expression.

     $ cat log                    
     stacktrace peak: 120 max: 70 allocs: 3 leaks: 50
      0xbbbe2bc3     lmdbg.c:101     log_stacktrace
      0xbbbe3498     lmdbg.c:456     realloc
      0x8049700      testme.c:902    func1
      0x8048788      testme.c:9      main
      0x80485b4      
      0x8048517      
     stacktrace peak: 100 max: 100 allocs: 1
      0xbbbe2bc3     lmdbg.c:101     log_stacktrace
      0xbbbe33bd     lmdbg.c:431     malloc
      0x8049654      testme.c:450    func2
      0x8048757      testme.c:7      main
      0x80485b4      
      0x8048517      
     $ lmdbg-grep 'leaks > 0' log 
     stacktrace peak: 120 max: 70 allocs: 3 leaks: 50
      0xbbbe2bc3     lmdbg.c:101     log_stacktrace
      0xbbbe3498     lmdbg.c:456     realloc
      0x8049700      testme.c:902    func1
      0x8048788      testme.c:9      main
      0x80485b4      
      0x8048517      
     $ lmdbg-grep '!leaks' log    
     stacktrace peak: 100 max: 100 allocs: 1
      0xbbbe2bc3     lmdbg.c:101     log_stacktrace
      0xbbbe33bd     lmdbg.c:431     malloc
      0x8049654      testme.c:450    func2
      0x8048757      testme.c:7      main
      0x80485b4      
      0x8048517      
     $ lmdbg-grep 'addrline ~ /\tfunc2/' log  
     stacktrace peak: 100 max: 100 allocs: 1
      0xbbbe2bc3     lmdbg.c:101     log_stacktrace
      0xbbbe33bd     lmdbg.c:431     malloc
      0x8049654      testme.c:450    func2
      0x8048757      testme.c:7      main
      0x80485b4      
      0x8048517      
     $ lmdbg-grep 'allocs > 1000 || peak > 1000000' log   
     $

- lmdbg-sort - see manual page.

====================================================================
                          QUICK GUIDE

lmdbg-run        --help
lmdbg-leaks      --help
lmdbg-sym        --help
lmdbg            --help
lmdbg-stat       --help
lmdbg-grep       --help

Sample of usage:

user@host$ lmdbg-run -o /tmp/PROG.log PROG ARG1 ARG2 ... ARGN
user@host$ lmdbg-leak < /tmp/PROG.log > /tmp/PROG.log2
user@host$ lmdbg-sym PROG \
   < /tmp/PROG.log2 > /tmp/PROG_LEAKS_SRC.log3

====================================================================
                AUTHOR: Aleksey Cheusov <vle@gmx.net>

====================================================================
                LICENSE: MIT

====================================================================
                      ACKNOWLEDGEMENTS

Idea to use full stacktraces was stolen from YAMD
  by Nate Eldredge

Idea to use gdb(1) for resolving symbols was stolen from ccmalloc
  by Armin Biere

Idea to use /proc/<pid>/maps was stolen from google-perftools
  by Google Inc. <opensource@google.com>

stacktrace(3) API is compatible with GNU libc's backtrace(3)


NEWS
======================================================================
Version 0.17.0, by Aleksey Cheusov, Thu, 24 Nov 2011 00:09:48 +0300

  lmdbg-sym:
    - Doesn't accept progname as an argument anymore.
      Use -P option instead.
    - option --with-so was replaced with -s.

  Long options have been removed from all programs.
  use standard single-letter POSIX-compatible options.

  SORRY FOR INCOMPATIBILITIES WITH PREVIOUS RELEASES.

                                    ----------------------------------

  lmdbg-stat, tests/test.sh:
    - Fixes for Solaris

  lmdbg-run:
    - options -T and -B were added to lmdbg-run.
    - option -n was added. It disables logging on startup.
      Sending SIGUSR1 signal to the process enables it.

  lmdbg:
    - -M option was added
    - -n option was added
    - -T and -B options were added
    - Regression tests were added for -M

  lmdbg-grep.1:
    - documentation for "bytes" variable

  If set, LMDBG_MODULES_AWK and LMDBG_STRIP_AWK environment variables
  are used as a path to AWK interpreter by "lmdbg-modules" and
  "lmdbg-strip" programs.

  More regression tests

======================================================================
Version 0.16.0, by Aleksey Cheusov, Thu, 16 Dec 2010 17:54:22 +0200

  Spelling corrections in man pages

  lmdbg-stat:
    - reimplemented in C. Now it uses libJudy library.
      See INSTALL for details.

  lmdbg-run:
    - NULL is *always* output as string "NULL"

======================================================================
Version 0.15.1, by Aleksey Cheusov, Thu,  7 Oct 2010 22:22:16 +0300

  FIX: On Linux/glibc systems lmdbg-run may fail (assert(3))
  because calloc(3) actually uses malloc(3). So, logging calloc(3)
  has been disabled.

  mk-configure >= 0.20 is required for building.

  Minor fixes and clean-ups.

======================================================================
Version 0.15.0, by Aleksey Cheusov, Wed, 26 May 2010 16:17:27 +0300

  Latest "mk-configure" is required for building.

  External tool "runawk" is needed for newly created "lmdbg-strip" and
  "lmdbg-modules" programs. http://sourceforge.net/projects/runawk/

  New utility "lmdbg-module" for enriching stacktraces given on input
  with information about module by set of rules specified by user.

  New utility "lmdbg-strip" for sripping stacktrace given on input.

  lmdbg-sort:
     - new sorting field "num"
     - minor fix in sorting order (Linux + gawk)
     - support for "info modulestat" input

  lmdbg-grep:
     - new variables to check: "op", "op_type" and "bytes".
     - new field to grep: "module".

  lmdbg-stat:
     - accepts "num: <num>" per-stacktrace information on input

  lmdbg:
     -ofilename is accepted and is equal to '-o filename'.

  lmdbg-run:
     - -ppipe is accepted and is equal to '-p pipe'.
     - by default 'num: <number>' tokens are output for each stacktrace.

  lmdbg-m2s, lmdbg-s2m:
    Internal singleline format has changed. As a result lmdbg-m2s|lmdbg-s2m
    pair works correctly with input that doesn't have real stacktraces.

  Fix in usage message for most lmdbg-* programs.

  Fixes for warnings messages produced by gcc.

  More regression tests.

======================================================================
Version 0.14.0, by Aleksey Cheusov, Tue, 13 Apr 2010 17:06:26 +0300

  New tools: lmdbg-stat, lmdbg-grep and lmdbg-sort for collecting and
  analysing statistical information about memory allocations.

  lmdbg-run: new options for lmdbg-run: -S and -M for generating
  shortened stacktraces.

  lmdbg-sym: new option -p for obtaining program name from lmdbg-run's
  output.

  fix: 'mkcmake test' remove its temporary files.

======================================================================
Version 0.13.0, by Aleksey Cheusov, Sun, 31 Jan 2010 18:40:05 +0200

  INSTALL: updated

  mk-configure>=0.14.0 is needed to build lmdbg.

  FIXED: 'make install' command installs test utilities from tests/
  subdirectory. Thanks to Mikolaj Golub for bug report.

  lmdbg-sym:
    - LMDBG uses very tricky features of GDB. As a result GDB may
      segfault on some platforms, e.g. gdb-6.4.90-debian fails on
      Debian/Etch.  In order to show users that GDB is buggy and needs
      to be fixed, an exit status of it is checked. For this purpose
      http://sf.net/projects/pipestatus/ is needed.
    - LMDBG_GDB environment variable sets the path to gdb program
      overiding path defined at build time. See scripts/Makefile.

  For better efficiency lmdbg-multi2single and lmdbg-single2multi
  internal scripts have been reimplemented in C and renamed to
  lmdbg-m2s and lmdbg-s2m respectively.

  Built-time AWK variable sets an AWK interpreter for awk script.
  See scripts/Makefile.

======================================================================
Version 0.12.0, by Aleksey Cheusov, Sun,  3 Jan 2010 14:05:15 +0200

  ADDED: support for calloc(3) and posix_memalign(3).

  Latest mk-configure is required for building.

  A target for testing lmdbg is again "test"! :-)

  Partial support for OpenBSD. Partial because I don't know how to
  detect where shared lib's code begins, there are no /proc/$$/maps.

  added: -a and -g options for "lmdbg" script.

  cleanups:
    - SHLIB_MAJOR/MINOR are used instead of dirty hacks.

  "make test" is now friendly with MAKEOBJDIR

  More regression tests.

======================================================================
Version 0.11.0, by Aleksey Cheusov, Sat, 25 Jul 2009 01:48:15 +0300

  Support for modern Linux/glibc where libraries are loaded
  to different regions of the virtual memory
  (See http://en.wikipedia.org/wiki/Address_space_layout_randomization).
  To handle this lmdbg-run adds the special "info section" information
  obtained from /proc/<pid>/maps available on NetBSD and Linux
  to the log file, and lmdbg-sym uses this information.

  fixed: enabling malloc logging in lmdbg-sym was an extreamly strange

stacktrace.c
/*
 * Copyright (c) 2007-2008 Aleksey Cheusov <vle@gmx.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#if HAVE_HEADER_EXECINFO_H
#include <execinfo.h>
/* GNU libc ? */
int stacktrace (void **buffer, int size)
{
	return backtrace (buffer, size);
}

#else
/* !HAVE_HEADER_EXECINFO_H (probably NetBSD/FreeBSD/Solaris etc.) */

#include <string.h>
#include <signal.h>
#include <setjmp.h>

static struct sigaction sigsegv_orig_handler;
static struct sigaction  sigbus_orig_handler;

static jmp_buf jmpbuf;

static void handler_sigfatal (int dummy)
{
	longjmp (jmpbuf,1);
}

static void set_sigfatal_handlers (void)
{
	struct sigaction sa;

	sa.sa_handler = handler_sigfatal;
	sigemptyset (&sa.sa_mask);
	sa.sa_flags = 0;

	sigaction (SIGSEGV, &sa, &sigsegv_orig_handler);
	sigaction (SIGBUS,  &sa,  &sigbus_orig_handler);
}

static void restore_sigfatal_handlers (void)
{
	sigaction (SIGSEGV, &sigsegv_orig_handler, NULL);
	sigaction (SIGBUS,  &sigbus_orig_handler, NULL);
}

#define one_return_address(x)                  \
		if (x >= size) break;                  \
		tb [x] = __builtin_return_address (x); \
		frame  = __builtin_frame_address (x); \
		if (!tb [x] || !frame){\
			tb [x] = 0; \
			break;\
		}

int stacktrace (void **tb, int size)
{
	unsigned i  = 0;
	void* frame = NULL;

	for (i=0; i < size; ++i){
		tb [i] = 0;
	}

	set_sigfatal_handlers ();

	if (!setjmp (jmpbuf)){
		while (1){
			one_return_address(0);
			one_return_address(1);
			one_return_address(2);
			one_return_address(3);
			one_return_address(4);
			one_return_address(5);
			one_return_address(6);
			one_return_address(7);
			one_return_address(8);
			one_return_address(9);
			one_return_address(10);
			one_return_address(11);
			one_return_address(12);
			one_return_address(13);
			one_return_address(14);
			one_return_address(15);
			one_return_address(16);
			one_return_address(17);
			one_return_address(18);
			one_return_address(19);
			one_return_address(20);
			one_return_address(21);
			one_return_address(22);
			one_return_address(23);
			one_return_address(24);
			one_return_address(25);
			one_return_address(26);
			one_return_address(27);
			one_return_address(28);
			one_return_address(29);
			one_return_address(30);
			one_return_address(31);
			one_return_address(32);
			one_return_address(33);
			one_return_address(34);
			one_return_address(35);
			one_return_address(36);
			one_return_address(37);
			one_return_address(38);
			one_return_address(39);
		}

		longjmp (jmpbuf, 2);
	}

	restore_sigfatal_handlers ();

	for (i=0; i < size; ++i){
		if (!tb [i])
			return i;
	}
	return size;
}
#endif /* HAVE_HEADER_EXECINFO_H */
